Content.Shared/Polymorph/Systems/SharedChameleonProjectorSystem.cs

```diff
@@ -1,67 +1,288 @@
-            || _whitelistSystem.IsBlacklistPass(comp.Blacklist, target);
+        return _whitelist.IsWhitelistFail(comp.Whitelist, target)
+            || _whitelist.IsBlacklistPass(comp.Blacklist, target);
     }
 
     /// <summary>
     /// On server, polymorphs the user into an entity and sets up the disguise.
     /// </summary>
-    public virtual void Disguise(ChameleonProjectorComponent comp, EntityUid user, EntityUid entity)
+    public void Disguise(Entity<ChameleonProjectorComponent> ent, EntityUid user, EntityUid entity)
     {
+        var proj = ent.Comp;
+
+        // no spawning prediction sorry
+        if (_net.IsClient)
+            return;
+
+        // reveal first to allow quick switching
+        TryReveal(user);
+
+        // add actions for controlling transform aspects
+        _actions.AddAction(user, ref proj.NoRotActionEntity, proj.NoRotAction, container: ent);
+        _actions.AddAction(user, ref proj.AnchorActionEntity, proj.AnchorAction, container: ent);
+
+        proj.Disguised = user;
+
+        var xform = Transform(user);
+        var disguise = Spawn(proj.DisguiseProto, xform.Coordinates);
+        var disguiseXform = Transform(disguise);
+        _xform.SetParent(disguise, disguiseXform, user, xform);
+
+        var disguised = AddComp<ChameleonDisguisedComponent>(user);
+        disguised.Disguise = disguise;
+        Dirty(user, disguised);
+
+        // make disguise look real (for simple things at least)
+        var meta = MetaData(entity);
+        _meta.SetEntityName(disguise, meta.EntityName);
+        _meta.SetEntityDescription(disguise, meta.EntityDescription);
+
+        var comp = EnsureComp<ChameleonDisguiseComponent>(disguise);
+        comp.User = user;
+        comp.Projector = ent;
+        comp.SourceEntity = entity;
+        comp.SourceProto = Prototype(entity)?.ID;
+        Dirty(disguise, comp);
+
+        // item disguises can be picked up to be revealed, also makes sure their examine size is correct
+        CopyComp<ItemComponent>((disguise, comp));
+
+        _appearance.CopyData(entity, disguise);
+
+        var mass = CompOrNull<PhysicsComponent>(entity)?.Mass ?? 0f;
+
+        // let the disguise die when its taken enough damage, which then transfers to the player
+        // health is proportional to mass, and capped to not be insane
+        if (TryComp<MobThresholdsComponent>(disguise, out var thresholds) && TryComp<MobThresholdsComponent>(user, out var userThresholds))
+        {
+            // cap disguise integrity at max health so you dont have to kill beforeif the player is of flesh and blood, cap max health to theirs
+            // so that when reverting damage scales 1:1 and not round removing
+            var playerMax = _mobThreshold.GetThresholdForState(user, MobState.Dead, userThresholds).Float();
+            var max = playerMax == 0f ? proj.MaxHealth : Math.Max(proj.MaxHealth, playerMax);
+            disguised.Integrity = max;
+        }
```

Comment:
Wouldnt it be simpler to just relay all the damage to the user? there must be a method somewhere that would respect the user armor and stuff and they can decide if the fact that an apple is visibly bleeding is something they should break disguise from and start running, either that or make the disguise be revealed on any damage

Reply:
legacy code i think it doesnt do anything anymore

Reply:
yeah you can only do damage by wide swinging left clicking the valid fuel tank does nothing

Reply:
probably good now